### ISSUE TITLE
Trigger only on textareas bigger than a specific size or disable for a specific site

### ISSUE THREAD

**Issue by simonweil** [glacambre/firenvim#248]
This is the first time I use firenvim and it looks amazing!

The problem I'm facing is that on some sites the textarea is so small that having nvim inside it makes it impossible to work...

Is it possible to configure firenvim to trigger only on textareas bigger than a specific size or disable for a specific site?

Thank you!

**Comment by glacambre** [comment]
You can't currently configure firenvim to only trigger on textareas bigger than a certain size, however you can configure it not to trigger on some websites. As explained in the [docs](https://github.com/glacambre/firenvim#configuring-the-browser-addon-behavior), you have to create an object named `firenvim_config` in your init.vim, create a key that matches the url of the page you want to disable firenvim on and the value of which is an object the `selector` value of which is an empty string. Basically,
```
let g:firenvim_config = {
    \ 'localSettings': {
        \ 'example.org': {
            \ 'selector': '',
        \ },
    \ }
\ }
```

Did you find this section of the documentation? If yes, how would you reword it to make it clearer that it is possible to prevent firenvim to automatically pop up on specific websites? And if not, what would you change to make this part of the documentation easier to find?

**Comment by simonweil** [comment]
Thank you.

I did read the docs but as I did it fast, I didn't comprehend that this way I can ignore a site.
I could understand it from:
```
let g:firenvim_config = {
    \ 'localSettings': {
        \ '.*': {
            \ 'selector': '',
            \ 'priority': 0,
        \ },
        \ 'github\.com': {
            \ 'selector': 'textarea',
            \ 'priority': 1,
        \ },
    \ }
\ }
```
But I need to think about it for that :)

Reading it again it is quite clear but maybe to make it clearer add just a specific example of removing a site as you sent me.

**Comment by simonweil** [comment]
But my problem is actually different.

Firenvim takes over intercom popups and they are always small, so it's not helpful.
Intercom popups appear in many sites, would it be possible to disable only intercom popups?

See this site for example: https://logz.io/pricing/

**Comment by glacambre** [comment]
You can use the `:not` selector to do that:
```
let g:firenvim_config = {
    \ 'localSettings': {
        \ '.*': {
            \ 'selector': '*:not(.intercom-composer) > textarea',
        \ },
    \ }
\ }
```
This CSS query means "select all textareas that are children of an element except for the ones that are children of an element the class of which is .intercom-composer".

Let me know if this works.

**Comment by simonweil** [comment]
I will check it out!
Thank you :)

Can I use multiple selectors?

**Comment by simonweil** [comment]
I managed to get it working like this:

```
let g:firenvim_config = {
    \ 'localSettings': {
        \ '.*': {
            \ 'selector': '*:not(.intercom-composer) > textarea',
            \ 'priority': 0,
        \ },
        \ 'docs\.google\.com': {
            \ 'selector': '',
            \ 'priority': 1,
        \ },
    \ }
\ }

```

The priority is important to explain as when I did the priority the opposite the goole did show the nvim editor.

I think you need more example configurations in to make it easier to setup.

Thank you again!

**Comment by glacambre** [comment]
Priorities are explained in the second paragraph of [Configuring the browser addon behavior](https://github.com/glacambre/firenvim#configuring-the-browser-addon-behavior): "When multiple patterns match a same URL, the pattern with the highest 'priority' value is used.". How would you reword it to make it clearer?

**Comment by BlueDrink9** [comment]
Is having a minimal size option for automatic triggering still being considered?

**Comment by glacambre** [comment]
The reason I want to avoid size limits is that they're pretty hard to 
implement, pixel density varies across screens so you have to measure 
things in points/characters but AFAIK there's no way to get a point's 
size and I can't measure the size of a character without creating a 
Neovim instance. On top of that there's already a lot of settings and 
the work required to ensure that they all work together as expected 
grows exponentially.

But if you describe a problem where this is the simplest or only 
solution, I would consider it yes :).

**Comment by BlueDrink9** [comment]
I was thinking if the instance would be <5 rows I wouldn't want to use it. What I can do is check for firenvim then just exit, but then I couldn't manually trigger in such a box. So an alternative might be a way to know if firenvim was automatically or manually triggered.

The motivation is just to make firenvim smarter about when it kicks in, without manually specifying rules for every site I visit. Personally I never need vim for small textboxes, especially since I only usually want to type and push enter (search boxes, chat apps) rather than edit.

**Comment by BlueDrink9** [comment]
Do you not create neovim instances for all textboxes anyway? Could you check row count there?

**Comment by glacambre** [comment]
No, I don't create a neovim instance per textbox, I only "pre-start" one 
and use it when the user selects a textbox. This pre-started instance 
doesn't have a GUI or anything, it just listens on a socket.

The problem is that the size of a row and column depend both on browser 
settings (which I don't have access to) and the `guifont` setting (which 
specifies both typeface and font size). So I have to create an iframe, 
attach to the preloaded instance, tell neovim I'm a gui, wait for it to 
send me Guifont settings, apply them to the content of the iframe and 
only then would I be able to tell whether the dimensions are ok or not :/

**Comment by BlueDrink9** [comment]
Ah yup, that doesn't sound feasible. What about a way to know if firenvim was triggered manually?

**Comment by alerque** [comment]
> I was thinking if the instance would be <5 rows I wouldn't want to use it.

Not to comment on this particular feature request, but maybe useful to you in setting up your environment. I find Firenvim useful even with 1 row. There are several things you can do to make it productive without taking up space. Disabling status bars if and only in in Firenvim or below a certain size _can_ be done. Also switching to the Firenvim provided command line instead of Neovim's saves another line and makes the integration with the browser a bit smoother. Combined all of these things mean the size of the input area is the least of my concerns. The only factors that make in unusable on some sites are unrelated to size. Your Millage May Vary™.

**Comment by BlueDrink9** [comment]
Yeah, I've done that, but I appreciate the suggestions. The space it takes up isn't the issue, except when other buttons get covered. It's more the extra overhead of entering insert mode, and entering command mode to manually save/quit. Even auto-entering insert mode and using ZZ, it's just niggly for small quick boxes, esp if doing it often like in a chat app (especially one that I haven't used before or not enough to have a config specific to it). Possibly also that in my head I'm not expecting firenvim in those smaller contexts, so it's a bit jarring.

**Comment by jbrockc** [comment]
> You can use the `:not` selector to do that:
> 
> ```
> let g:firenvim_config = {
>     \ 'localSettings': {
>         \ '.*': {
>             \ 'selector': '*:not(.intercom-composer) > textarea',
>         \ },
>     \ }
> \ }
> ```
> 
> This CSS query means "select all textareas that are children of an element except for the ones that are children of an element the class of which is `intercom-composer`".
> 
> Let me know if this works.

This solution worked beautifully for me. Outlook Web App changed, and Firenvim was triggering in the To, Cc, and Bcc fields. The `<div>` has `role="textbox"`. Now it only triggers in the message field. Thank you!
